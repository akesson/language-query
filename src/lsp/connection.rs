use anyhow::Result;
use async_trait::async_trait;
use std::path::Path;

#[async_trait]
pub trait LspConnection: Send + Sync {
    async fn hover(&self, file: &Path, line: u32, symbol: &str) -> Result<Option<String>>;
    async fn implementation(&self, file: &Path, line: u32, symbol: &str) -> Result<Option<String>>;
    async fn references(&self, file: &Path, line: u32, symbol: &str) -> Result<Vec<String>>;
    async fn resolve_symbol(&self, file: &Path, symbol: &str) -> Result<Option<String>>;
}