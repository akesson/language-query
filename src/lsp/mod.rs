pub mod connection;
pub mod rust_analyzer;

pub use connection::LspConnection;
pub use rust_analyzer::RustAnalyzerConnection;