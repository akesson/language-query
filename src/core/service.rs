use std::path::{Path, PathBuf};
use anyhow::{Result, bail};

use crate::lsp::{LspConnection, RustAnalyzerConnection};

pub struct LanguageQueryService {
    lsp: Box<dyn LspConnection>,
    workspace: PathBuf,
}

impl LanguageQueryService {
    pub async fn new(workspace: &Path) -> Result<Self> {
        // Check if we have a Rust project
        if workspace.join("Cargo.toml").exists() {
            let lsp = Box::new(RustAnalyzerConnection::new(workspace).await?);
            Ok(Self {
                lsp,
                workspace: workspace.to_path_buf(),
            })
        } else {
            bail!("Not a Rust project (no Cargo.toml found). Only Rust projects are currently supported.")
        }
    }
    
    pub fn workspace_path(&self) -> &Path {
        &self.workspace
    }
    
    pub async fn get_docs(&self, file: &Path, line: u32, symbol: &str) -> Result<Option<String>> {
        self.lsp.hover(file, line, symbol).await
    }
    
    pub async fn get_impl(&self, file: &Path, line: u32, symbol: &str) -> Result<Option<String>> {
        self.lsp.implementation(file, line, symbol).await
    }
    
    pub async fn get_refs(&self, file: &Path, line: u32, symbol: &str) -> Result<Vec<String>> {
        self.lsp.references(file, line, symbol).await
    }
    
    pub async fn resolve_symbol(&self, file: &Path, symbol: &str) -> Result<Option<String>> {
        self.lsp.resolve_symbol(file, symbol).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;
    
    fn ensure_rust_analyzer() {
        // Ensure rust-analyzer is available
        std::process::Command::new("rust-analyzer")
            .arg("--version")
            .output()
            .expect("rust-analyzer not found. Please install rust-analyzer to run tests.");
    }
    
    async fn create_test_project() -> Result<(TempDir, PathBuf)> {
        let temp_dir = TempDir::new()?;
        
        // Create a minimal Cargo.toml
        std::fs::write(
            temp_dir.path().join("Cargo.toml"),
            r#"[package]
name = "test_project"
version = "0.1.0"
edition = "2021"
"#
        )?;
        
        // Create src directory
        std::fs::create_dir_all(temp_dir.path().join("src"))?;
        
        // Create a simple lib.rs
        let lib_file = temp_dir.path().join("src/lib.rs");
        std::fs::write(&lib_file, r#"//! Test library

/// A test struct
pub struct TestStruct {
    /// The value field
    pub value: String,
}

impl TestStruct {
    /// Creates a new TestStruct
    pub fn new(value: String) -> Self {
        Self { value }
    }
}

/// Test function using TestStruct
pub fn use_test_struct(ts: &TestStruct) -> &str {
    &ts.value
}

use std::collections::HashMap;

/// Creates a test map
pub fn create_map() -> HashMap<String, i32> {
    let mut map = HashMap::new();
    map.insert("test".to_string(), 42);
    map
}
"#)?;
        
        Ok((temp_dir, lib_file))
    }
    
    fn redact_temp_path(content: &str, temp_path: &Path) -> String {
        content.replace(&temp_path.to_string_lossy().to_string(), "[TEMP_DIR]")
    }
    
    #[tokio::test]
    async fn test_service_creation() {
        ensure_rust_analyzer();
        
        let (temp_dir, _) = create_test_project().await.unwrap();
        
        // Test that we can create a service
        let service = LanguageQueryService::new(temp_dir.path()).await.unwrap();
        assert_eq!(service.workspace_path(), temp_dir.path());
    }
    
    #[tokio::test]
    async fn test_non_rust_project_fails() {
        ensure_rust_analyzer();
        
        let temp_dir = TempDir::new().unwrap();
        
        // Should fail without Cargo.toml
        let result = LanguageQueryService::new(temp_dir.path()).await;
        assert!(result.is_err());
        let err_msg = result.err().unwrap().to_string();
        assert!(err_msg.contains("Not a Rust project"));
    }
    
    #[tokio::test]
    async fn test_docs_command() {
        ensure_rust_analyzer();
        
        let (temp_dir, lib_file) = create_test_project().await.unwrap();
        let service = LanguageQueryService::new(temp_dir.path()).await.unwrap();
        
        // No need to wait - the service waits for ready internally
        
        // Get docs for TestStruct
        let result = service.get_docs(&lib_file, 4, "TestStruct").await;
        
        match result {
            Ok(Some(docs)) => {
                // Redact the temp directory path
                let redacted = redact_temp_path(&docs, temp_dir.path());
                insta::assert_snapshot!("test_docs_command", redacted);
            }
            Ok(None) => panic!("Expected documentation for TestStruct"),
            Err(e) => panic!("Failed to get docs: {}", e),
        }
    }
    
    #[tokio::test]
    async fn test_impl_command() {
        ensure_rust_analyzer();
        
        let (temp_dir, lib_file) = create_test_project().await.unwrap();
        let service = LanguageQueryService::new(temp_dir.path()).await.unwrap();
        
        // Try to get implementation of TestStruct at the struct definition
        let result = service.get_impl(&lib_file, 4, "TestStruct").await.unwrap();
        
        if let Some(implementation) = result {
            let redacted = redact_temp_path(&implementation, temp_dir.path());
            insta::assert_snapshot!("test_impl_command", redacted);
        } else {
            // It's OK if go-to-definition returns None for struct definition
            println!("No implementation found (acceptable for struct definition)");
        }
    }
    
    #[tokio::test]
    async fn test_refs_command() {
        ensure_rust_analyzer();
        
        let (temp_dir, lib_file) = create_test_project().await.unwrap();
        let service = LanguageQueryService::new(temp_dir.path()).await.unwrap();
        
        // Find references to TestStruct
        let result = service.get_refs(&lib_file, 4, "TestStruct").await;
        
        match result {
            Ok(refs) => {
                assert!(!refs.is_empty(), "Expected at least one reference to TestStruct");
                
                // Redact temp paths in all references
                let redacted_refs: Vec<String> = refs.iter()
                    .map(|r| redact_temp_path(r, temp_dir.path()))
                    .collect();
                
                insta::assert_snapshot!("test_refs_command", redacted_refs.join("\n"));
            }
            Err(e) => panic!("Failed to get references: {}", e),
        }
    }
    
    #[tokio::test]
    async fn test_resolve_command() {
        ensure_rust_analyzer();
        
        let (temp_dir, lib_file) = create_test_project().await.unwrap();
        let service = LanguageQueryService::new(temp_dir.path()).await.unwrap();
        
        // Resolve HashMap
        let result = service.resolve_symbol(&lib_file, "HashMap").await.unwrap();
        
        if let Some(resolved) = result {
            let redacted = redact_temp_path(&resolved, temp_dir.path());
            insta::assert_snapshot!("test_resolve_command", redacted);
        } else {
            panic!("Expected to resolve HashMap");
        }
    }
}