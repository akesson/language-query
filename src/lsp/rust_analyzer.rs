use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::sync::Arc;
use std::ops::ControlFlow;
use std::collections::HashSet;
use std::time::Duration;

use anyhow::{Result, Context, bail};
use async_lsp::concurrency::ConcurrencyLayer;
use async_lsp::panic::CatchUnwindLayer;
use async_lsp::server::LifecycleLayer;
use async_lsp::tracing::TracingLayer;
use async_lsp::{LanguageServer, MainLoop, ServerSocket};
use async_lsp::router::Router;
use async_process::Command;
use async_trait::async_trait;
use lsp_types::{
    ClientCapabilities, DidOpenTextDocumentParams, GotoDefinitionParams,
    GotoDefinitionResponse, Hover, HoverClientCapabilities, HoverParams, InitializeParams,
    InitializedParams, MarkupKind, Position, ReferenceContext, ReferenceParams,
    TextDocumentClientCapabilities, TextDocumentIdentifier, TextDocumentItem,
    TextDocumentPositionParams, WindowClientCapabilities, WorkDoneProgressParams,
    WorkspaceFolder,
};
use tokio::sync::{Mutex, RwLock};
use tokio::task;
use tower::ServiceBuilder;
use tracing::{info, error};

use crate::lsp::LspConnection;

pub struct RustAnalyzerConnection {
    workspace: PathBuf,
    server: Arc<Mutex<ServerSocket>>,
    opened_files: Arc<Mutex<HashSet<PathBuf>>>,
    is_ready: Arc<RwLock<bool>>,
    #[allow(dead_code)]
    _mainloop_handle: tokio::task::JoinHandle<()>,
    #[allow(dead_code)]
    _child: async_process::Child,
}

impl RustAnalyzerConnection {
    pub async fn new(workspace: &Path) -> Result<Self> {
        let rust_analyzer_path = find_rust_analyzer()?;
        
        info!("Starting rust-analyzer at: {:?}", rust_analyzer_path);
        
        let mut child = Command::new(&rust_analyzer_path)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .current_dir(workspace)
            .kill_on_drop(true)
            .spawn()
            .context("Failed to spawn rust-analyzer")?;
        
        let stdout = child.stdout.take().unwrap();
        let stdin = child.stdin.take().unwrap();
        
        // Create the main loop for LSP communication
        let (mainloop, server) = MainLoop::new_client(|_server| {
            ServiceBuilder::new()
                .layer(TracingLayer::default())
                .layer(LifecycleLayer::default())
                .layer(CatchUnwindLayer::default())
                .layer(ConcurrencyLayer::default())
                .service(ClientState::new_router())
        });
        
        let server = Arc::new(Mutex::new(server));
        
        // Run the main loop in a background task
        let mainloop_handle = task::spawn(async move {
            if let Err(e) = mainloop.run_buffered(stdout, stdin).await {
                error!("Language server mainloop error: {}", e);
            }
        });
        
        let is_ready = Arc::new(RwLock::new(false));
        
        let connection = Self {
            workspace: workspace.to_path_buf(),
            server,
            opened_files: Arc::new(Mutex::new(HashSet::new())),
            is_ready: is_ready.clone(),
            _mainloop_handle: mainloop_handle,
            _child: child,
        };
        
        // Initialize the LSP server
        connection.initialize().await?;
        
        // Wait for the server to be ready
        connection.wait_until_ready().await?;
        
        Ok(connection)
    }
    
    async fn initialize(&self) -> Result<()> {
        info!("Initializing rust-analyzer for workspace: {:?}", self.workspace);
        
        let initialize_params = InitializeParams {
            workspace_folders: Some(vec![WorkspaceFolder {
                uri: url::Url::from_file_path(&self.workspace)
                    .map_err(|_| anyhow::anyhow!("Invalid workspace path: {:?}", self.workspace))?,
                name: self.workspace.file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .to_string(),
            }]),
            capabilities: ClientCapabilities {
                text_document: Some(TextDocumentClientCapabilities {
                    hover: Some(HoverClientCapabilities {
                        content_format: Some(vec![MarkupKind::Markdown]),
                        ..Default::default()
                    }),
                    ..Default::default()
                }),
                window: Some(WindowClientCapabilities {
                    work_done_progress: Some(true),
                    ..Default::default()
                }),
                ..Default::default()
            },
            ..Default::default()
        };
        
        let mut server = self.server.lock().await;
        let _response = server.initialize(initialize_params).await?;
        server.initialized(InitializedParams {})?;
        
        Ok(())
    }
    
    async fn wait_until_ready(&self) -> Result<()> {
        info!("Waiting for rust-analyzer to be ready...");
        
        // For rust-analyzer, we need to wait for it to index the project
        // The best approach is to wait a reasonable amount of time based on project size
        // For test projects, 2-3 seconds should be enough
        
        // In a production system, we could:
        // 1. Use workspace/symbol requests to check if symbols are indexed
        // 2. Monitor progress notifications from the server
        // 3. Use experimental/serverStatus if available
        
        // For now, use a simple time-based approach
        let wait_time = if cfg!(test) {
            Duration::from_secs(3) // Longer wait in tests
        } else {
            Duration::from_secs(2) // Normal wait
        };
        
        info!("Waiting {:?} for rust-analyzer to initialize...", wait_time);
        tokio::time::sleep(wait_time).await;
        
        *self.is_ready.write().await = true;
        info!("rust-analyzer marked as ready");
        
        Ok(())
    }
    
    async fn ensure_ready(&self) -> Result<()> {
        if !*self.is_ready.read().await {
            bail!("LSP server is not ready yet");
        }
        Ok(())
    }
    
    async fn open_file(&self, file: &Path) -> Result<()> {
        // Make the path absolute if it's relative
        let absolute_path = if file.is_absolute() {
            file.to_path_buf()
        } else {
            std::env::current_dir()?.join(file)
        };
        
        let canonical_path = absolute_path.canonicalize()
            .unwrap_or_else(|_| absolute_path.clone());
        
        // Check if file is already open
        let mut opened = self.opened_files.lock().await;
        if opened.contains(&canonical_path) {
            return Ok(());
        }
        
        let uri = url::Url::from_file_path(&canonical_path)
            .map_err(|_| {
                error!("Failed to create URI from path: {:?}", canonical_path);
                anyhow::anyhow!("Invalid file path: {:?}", canonical_path)
            })?;
        
        let contents = tokio::fs::read_to_string(&canonical_path)
            .await
            .context("Failed to read file")?;
        
        let params = DidOpenTextDocumentParams {
            text_document: TextDocumentItem {
                uri,
                language_id: "rust".to_string(),
                version: 0,
                text: contents,
            },
        };
        
        let mut server = self.server.lock().await;
        server.did_open(params)?;
        
        // Mark file as opened
        opened.insert(canonical_path);
        
        Ok(())
    }
    
    /// Find the position of a symbol in a file starting from the given line
    async fn find_symbol_position(&self, file: &Path, line: u32, symbol: &str) -> Result<Position> {
        let contents = tokio::fs::read_to_string(file)
            .await
            .context("Failed to read file")?;
        
        let lines: Vec<&str> = contents.lines().collect();
        
        // Convert 1-based line to 0-based
        let line_index = (line as usize).saturating_sub(1);
        
        if line_index >= lines.len() {
            bail!("Line {} is out of bounds (file has {} lines)", line, lines.len());
        }
        
        // Search for the symbol in the specified line
        if let Some(char_index) = lines[line_index].find(symbol) {
            Ok(Position {
                line: line_index as u32,
                character: char_index as u32,
            })
        } else {
            // If not found on the exact line, search nearby lines
            for offset in 1..=2 {
                // Check lines before
                if line_index >= offset {
                    let check_line = line_index - offset;
                    if let Some(char_index) = lines[check_line].find(symbol) {
                        return Ok(Position {
                            line: check_line as u32,
                            character: char_index as u32,
                        });
                    }
                }
                
                // Check lines after
                let check_line = line_index + offset;
                if check_line < lines.len() {
                    if let Some(char_index) = lines[check_line].find(symbol) {
                        return Ok(Position {
                            line: check_line as u32,
                            character: char_index as u32,
                        });
                    }
                }
            }
            
            bail!("Symbol '{}' not found near line {}", symbol, line);
        }
    }
}

#[async_trait]
impl LspConnection for RustAnalyzerConnection {
    async fn hover(&self, file: &Path, line: u32, symbol: &str) -> Result<Option<String>> {
        // Ensure server is ready
        self.ensure_ready().await?;
        
        // Ensure file is open
        self.open_file(file).await?;
        
        let position = self.find_symbol_position(file, line, symbol).await?;
        
        // Make the path absolute if it's relative
        let absolute_path = if file.is_absolute() {
            file.to_path_buf()
        } else {
            std::env::current_dir()?.join(file)
        };
        
        let uri = url::Url::from_file_path(&absolute_path)
            .map_err(|_| anyhow::anyhow!("Invalid file path: {:?}", absolute_path))?;
        
        let params = HoverParams {
            text_document_position_params: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier { uri },
                position,
            },
            work_done_progress_params: WorkDoneProgressParams {
                work_done_token: None,
            },
        };
        
        // Retry logic for content modified errors
        let mut attempts = 0;
        loop {
            let mut server = self.server.lock().await;
            match server.hover(params.clone()).await {
                Ok(response) => {
                    if let Some(hover) = response {
                        let content = format_hover_content(&hover);
                        return Ok(Some(content));
                    } else {
                        return Ok(None);
                    }
                }
                Err(e) if e.to_string().contains("content modified") && attempts < 3 => {
                    drop(server); // Release the lock
                    attempts += 1;
                    info!("Retrying hover request due to content modified error (attempt {})", attempts);
                    tokio::time::sleep(Duration::from_millis(500)).await;
                }
                Err(e) => return Err(e.into()),
            }
        }
    }
    
    async fn implementation(&self, file: &Path, line: u32, symbol: &str) -> Result<Option<String>> {
        // Ensure server is ready
        self.ensure_ready().await?;
        
        // Ensure file is open
        self.open_file(file).await?;
        
        let position = self.find_symbol_position(file, line, symbol).await?;
        
        // Make the path absolute if it's relative
        let absolute_path = if file.is_absolute() {
            file.to_path_buf()
        } else {
            std::env::current_dir()?.join(file)
        };
        
        let uri = url::Url::from_file_path(&absolute_path)
            .map_err(|_| anyhow::anyhow!("Invalid file path: {:?}", absolute_path))?;
        
        let params = GotoDefinitionParams {
            text_document_position_params: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier { uri },
                position,
            },
            work_done_progress_params: WorkDoneProgressParams {
                work_done_token: None,
            },
            partial_result_params: Default::default(),
        };
        
        let mut server = self.server.lock().await;
        
        // Try to get the definition first
        let response = server.definition(params).await?;
        
        if let Some(GotoDefinitionResponse::Scalar(location)) = response {
            // Read the implementation from the file
            let impl_path = location.uri.to_file_path()
                .map_err(|_| anyhow::anyhow!("Invalid URI"))?;
            let contents = tokio::fs::read_to_string(&impl_path).await?;
            let lines: Vec<&str> = contents.lines().collect();
            
            let start_line = location.range.start.line as usize;
            let end_line = location.range.end.line as usize;
            
            // Extract more context around the definition
            let context_start = start_line.saturating_sub(1);
            let context_end = (end_line + 10).min(lines.len() - 1);
            
            let impl_lines: Vec<String> = lines[context_start..=context_end]
                .iter()
                .map(|s| s.to_string())
                .collect();
            
            let relative_path = impl_path.strip_prefix(&self.workspace).unwrap_or(&impl_path);
            
            Ok(Some(format!(
                "{}:{}:{}:\n```rust\n{}\n```",
                relative_path.display(),
                location.range.start.line + 1,
                location.range.end.line + 1,
                impl_lines.join("\n")
            )))
        } else {
            Ok(None)
        }
    }
    
    async fn references(&self, file: &Path, line: u32, symbol: &str) -> Result<Vec<String>> {
        // Ensure server is ready
        self.ensure_ready().await?;
        
        // Ensure file is open
        self.open_file(file).await?;
        
        let position = self.find_symbol_position(file, line, symbol).await?;
        
        // Make the path absolute if it's relative
        let absolute_path = if file.is_absolute() {
            file.to_path_buf()
        } else {
            std::env::current_dir()?.join(file)
        };
        
        let uri = url::Url::from_file_path(&absolute_path)
            .map_err(|_| anyhow::anyhow!("Invalid file path: {:?}", absolute_path))?;
        
        let params = ReferenceParams {
            text_document_position: TextDocumentPositionParams {
                text_document: TextDocumentIdentifier { uri },
                position,
            },
            work_done_progress_params: WorkDoneProgressParams {
                work_done_token: None,
            },
            partial_result_params: Default::default(),
            context: ReferenceContext {
                include_declaration: false,
            },
        };
        
        // Retry logic for content modified errors
        let mut attempts = 0;
        loop {
            let mut server = self.server.lock().await;
            match server.references(params.clone()).await {
                Ok(response) => {
                    drop(server); // Release lock before doing I/O
                    
                    let mut results = Vec::new();
                    
                    if let Some(locations) = response {
                        for location in locations {
                            let ref_path = location.uri.to_file_path()
                                .map_err(|_| anyhow::anyhow!("Invalid URI"))?;
                            
                            let relative_path = ref_path.strip_prefix(&self.workspace).unwrap_or(&ref_path);
                            
                            // Read the line to show context
                            let contents = tokio::fs::read_to_string(&ref_path).await?;
                            let lines: Vec<&str> = contents.lines().collect();
                            let line_num = location.range.start.line as usize;
                            
                            if line_num < lines.len() {
                                let line_content = lines[line_num].trim();
                                results.push(format!(
                                    "{}:{}: {}",
                                    relative_path.display(),
                                    line_num + 1,
                                    line_content
                                ));
                            }
                        }
                    }
                    
                    return Ok(results);
                }
                Err(e) if e.to_string().contains("content modified") && attempts < 3 => {
                    drop(server); // Release the lock
                    attempts += 1;
                    info!("Retrying references request due to content modified error (attempt {})", attempts);
                    tokio::time::sleep(Duration::from_millis(500)).await;
                }
                Err(e) => return Err(e.into()),
            }
        }
    }
    
    async fn resolve_symbol(&self, file: &Path, symbol: &str) -> Result<Option<String>> {
        // Ensure server is ready
        self.ensure_ready().await?;
        
        // For symbol resolution, we'll use hover at the first occurrence
        info!("Attempting to resolve symbol '{}' in file: {:?}", symbol, file);
        let contents = tokio::fs::read_to_string(file)
            .await
            .with_context(|| format!("Failed to read file: {:?}", file))?;
        
        // Find the first occurrence of the symbol
        for (line_num, line) in contents.lines().enumerate() {
            if line.contains(symbol) {
                if let Ok(Some(hover)) = self.hover(file, (line_num + 1) as u32, symbol).await {
                    return Ok(Some(format!(
                        "Found symbol `{}` in {}:\n\n{}",
                        symbol,
                        file.file_name().unwrap_or_default().to_string_lossy(),
                        hover
                    )));
                }
            }
        }
        
        Ok(None)
    }
}

fn find_rust_analyzer() -> Result<PathBuf> {
    // Try to find rust-analyzer in PATH
    if let Ok(output) = std::process::Command::new("which")
        .arg("rust-analyzer")
        .output()
    {
        if output.status.success() {
            let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
            return Ok(PathBuf::from(path));
        }
    }
    
    // Common locations to check
    let common_paths = [
        "/usr/local/bin/rust-analyzer",
        "/usr/bin/rust-analyzer",
        "/opt/homebrew/bin/rust-analyzer",
        "~/.cargo/bin/rust-analyzer",
    ];
    
    for path in &common_paths {
        let expanded = shellexpand::tilde(path);
        let path = PathBuf::from(expanded.as_ref());
        if path.exists() {
            return Ok(path);
        }
    }
    
    bail!("Could not find rust-analyzer. Please ensure it is installed and in your PATH.")
}

fn format_hover_content(hover: &Hover) -> String {
    use lsp_types::HoverContents;
    use lsp_types::MarkedString;
    
    match &hover.contents {
        HoverContents::Scalar(marked) => match marked {
            MarkedString::String(s) => s.clone(),
            MarkedString::LanguageString(ls) => {
                format!("```{}\n{}\n```", ls.language, ls.value)
            }
        },
        HoverContents::Array(marked_strings) => {
            marked_strings
                .iter()
                .map(|ms| match ms {
                    MarkedString::String(s) => s.clone(),
                    MarkedString::LanguageString(ls) => {
                        format!("```{}\n{}\n```", ls.language, ls.value)
                    }
                })
                .collect::<Vec<_>>()
                .join("\n\n")
        }
        HoverContents::Markup(markup) => markup.value.clone(),
    }
}

// Minimal client state to handle LSP notifications
#[derive(Clone)]
struct ClientState;

impl ClientState {
    fn new_router() -> Router<Self> {
        let mut router = Router::new(ClientState);
        
        router.notification::<lsp_types::notification::ShowMessage>(|_state, _params| {
            ControlFlow::Continue(())
        });
        
        router.notification::<lsp_types::notification::LogMessage>(|_state, _params| {
            ControlFlow::Continue(())
        });
        
        router.notification::<lsp_types::notification::PublishDiagnostics>(|_state, _params| {
            ControlFlow::Continue(())
        });
        
        router.notification::<lsp_types::notification::Progress>(|_state, _params| {
            ControlFlow::Continue(())
        });
        
        router
    }
}

impl async_lsp::LanguageClient for ClientState {
    type Error = async_lsp::Error;
    type NotifyResult = ControlFlow<Result<(), Self::Error>>;

    fn progress(&mut self, _params: lsp_types::ProgressParams) -> Self::NotifyResult {
        ControlFlow::Continue(())
    }

    fn publish_diagnostics(&mut self, _params: lsp_types::PublishDiagnosticsParams) -> Self::NotifyResult {
        ControlFlow::Continue(())
    }

    fn log_message(&mut self, _params: lsp_types::LogMessageParams) -> Self::NotifyResult {
        ControlFlow::Continue(())
    }

    fn show_message(&mut self, _params: lsp_types::ShowMessageParams) -> Self::NotifyResult {
        ControlFlow::Continue(())
    }
}