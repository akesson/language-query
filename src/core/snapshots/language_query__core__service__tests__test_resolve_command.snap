---
source: src/core/service.rs
expression: redacted
---
Found symbol `HashMap` in lib.rs:


```rust
std::collections::hash::map
```

```rust
pub struct HashMap<K, V, S = RandomState> {
    base: HashMap<K, V, S>,
}
```

---

A [hash map] implemented with quadratic probing and SIMD lookup.

By default, `HashMap` uses a hashing algorithm selected to provide
resistance against HashDoS attacks. The algorithm is randomly seeded, and a
reasonable best-effort is made to generate this seed from a high quality,
secure source of randomness provided by the host without blocking the
program. Because of this, the randomness of the seed depends on the output
quality of the system's random number coroutine when the seed is created.
In particular, seeds generated when the system's entropy pool is abnormally
low such as during system boot may be of a lower quality.

The default hashing algorithm is currently SipHash 1-3, though this is
subject to change at any point in the future. While its performance is very
competitive for medium sized keys, other hashing algorithms will outperform
it for small keys such as integers as well as large keys such as long
strings, though those algorithms will typically *not* protect against
attacks such as HashDoS.

The hashing algorithm can be replaced on a per-`HashMap` basis using the
[`default`], [`with_hasher`], and [`with_capacity_and_hasher`] methods.
There are many alternative [hashing algorithms available on crates.io](https://crates.io/keywords/hasher).

It is required that the keys implement the [`Eq`](https://doc.rust-lang.org/stable/core/cmp/trait.Eq.html) and [`Hash`](https://doc.rust-lang.org/stable/core/hash/trait.Hash.html) traits, although
this can frequently be achieved by using `#[derive(PartialEq, Eq, Hash)]`.
If you implement these yourself, it is important that the following
property holds:

```text
k1 == k2 -> hash(k1) == hash(k2)
```

In other words, if two keys are equal, their hashes must be equal.
Violating this property is a logic error.

It is also a logic error for a key to be modified in such a way that the key's
hash, as determined by the [`Hash`](https://doc.rust-lang.org/stable/core/hash/trait.Hash.html) trait, or its equality, as determined by
the [`Eq`](https://doc.rust-lang.org/stable/core/cmp/trait.Eq.html) trait, changes while it is in the map. This is normally only
possible through [`Cell`], [`RefCell`], global state, I/O, or unsafe code.

The behavior resulting from either logic error is not specified, but will
be encapsulated to the `HashMap` that observed the logic error and not
result in undefined behavior. This could include panics, incorrect results,
aborts, memory leaks, and non-termination.

The hash table implementation is a Rust port of Google's [SwissTable](https://abseil.io/blog/20180927-swisstables).
The original C++ version of SwissTable can be found [here](https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/raw_hash_set.h), and this
[CppCon talk](https://www.youtube.com/watch?v=ncHmEUmJZf4) gives an overview of how the algorithm works.

# Examples

```rust
use std::collections::HashMap;

// Type inference lets us omit an explicit type signature (which
// would be `HashMap<String, String>` in this example).
let mut book_reviews = HashMap::new();

// Review some books.
book_reviews.insert(
    "Adventures of Huckleberry Finn".to_string(),
    "My favorite book.".to_string(),
);
book_reviews.insert(
    "Grimms' Fairy Tales".to_string(),
    "Masterpiece.".to_string(),
);
book_reviews.insert(
    "Pride and Prejudice".to_string(),
    "Very enjoyable.".to_string(),
);
book_reviews.insert(
    "The Adventures of Sherlock Holmes".to_string(),
    "Eye lyked it alot.".to_string(),
);

// Check for a specific one.
// When collections store owned values (String), they can still be
// queried using references (&str).
if !book_reviews.contains_key("Les Misérables") {
    println!("We've got {} reviews, but Les Misérables ain't one.",
             book_reviews.len());
}

// oops, this review has a lot of spelling mistakes, let's delete it.
book_reviews.remove("The Adventures of Sherlock Holmes");

// Look up the values associated with some keys.
let to_find = ["Pride and Prejudice", "Alice's Adventure in Wonderland"];
for &book in &to_find {
    match book_reviews.get(book) {
        Some(review) => println!("{book}: {review}"),
        None => println!("{book} is unreviewed.")
    }
}

// Look up the value for a key (will panic if the key is not found).
println!("Review for Jane: {}", book_reviews["Pride and Prejudice"]);

// Iterate over everything.
for (book, review) in &book_reviews {
    println!("{book}: \"{review}\"");
}
```

A `HashMap` with a known list of items can be initialized from an array:

```rust
use std::collections::HashMap;

let solar_distance = HashMap::from([
    ("Mercury", 0.4),
    ("Venus", 0.7),
    ("Earth", 1.0),
    ("Mars", 1.5),
]);
```

`HashMap` implements an [`Entry` API](https://doc.rust-lang.org/stable/std/collections/hash/map/struct.HashMap.html#method.entry), which allows
for complex methods of getting, setting, updating and removing keys and
their values:

```rust
use std::collections::HashMap;

// type inference lets us omit an explicit type signature (which
// would be `HashMap<&str, u8>` in this example).
let mut player_stats = HashMap::new();

fn random_stat_buff() -> u8 {
    // could actually return some random value here - let's just return
    // some fixed value for now
    42
}

// insert a key only if it doesn't already exist
player_stats.entry("health").or_insert(100);

// insert a key using a function that provides a new value only if it
// doesn't already exist
player_stats.entry("defence").or_insert_with(random_stat_buff);

// update a key, guarding against the key possibly not being set
let stat = player_stats.entry("attack").or_insert(100);
*stat += random_stat_buff();

// modify an entry before an insert with in-place mutation
player_stats.entry("mana").and_modify(|mana| *mana += 200).or_insert(100);
```

The easiest way to use `HashMap` with a custom key type is to derive [`Eq`](https://doc.rust-lang.org/stable/core/cmp/trait.Eq.html) and [`Hash`](https://doc.rust-lang.org/stable/core/hash/trait.Hash.html).
We must also derive [`PartialEq`](https://doc.rust-lang.org/stable/core/cmp/trait.PartialEq.html).

```rust
use std::collections::HashMap;

#[derive(Hash, Eq, PartialEq, Debug)]
struct Viking {
    name: String,
    country: String,
}

impl Viking {
    /// Creates a new Viking.
    fn new(name: &str, country: &str) -> Viking {
        Viking { name: name.to_string(), country: country.to_string() }
    }
}

// Use a HashMap to store the vikings' health points.
let vikings = HashMap::from([
    (Viking::new("Einar", "Norway"), 25),
    (Viking::new("Olaf", "Denmark"), 24),
    (Viking::new("Harald", "Iceland"), 12),
]);

// Use derived implementation to print the status of the vikings.
for (viking, health) in &vikings {
    println!("{viking:?} has {health} hp");
}
```

# Usage in `const` and `static`

As explained above, `HashMap` is randomly seeded: each `HashMap` instance uses a different seed,
which means that `HashMap::new` cannot be used in const context. To construct a `HashMap` in the
initializer of a `const` or `static` item, you will have to use a different hasher that does not
involve a random seed, as demonstrated in the following example. **A `HashMap` constructed this
way is not resistant against HashDoS!**

```rust
use std::collections::HashMap;
use std::hash::{BuildHasherDefault, DefaultHasher};
use std::sync::Mutex;

const EMPTY_MAP: HashMap<String, Vec<i32>, BuildHasherDefault<DefaultHasher>> =
    HashMap::with_hasher(BuildHasherDefault::new());
static MAP: Mutex<HashMap<String, Vec<i32>, BuildHasherDefault<DefaultHasher>>> =
    Mutex::new(HashMap::with_hasher(BuildHasherDefault::new()));
```
